_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+
+The next simplest approach is to pass in a unique string to the assert which
+provides context for the developer about what the error was and where. We can
+then search for this error message within the codebase and figure out at which
+location the assert took place.
+
+```c
+#define MY_ASSERT(expr, msg) \
+  do {                                          \
+    if (!(expr)) {                              \
+      my_assert(msg);                           \
+    }                                           \
+  } while (0)
+```
+
+```c
+void my_assert(const char *msg) {
+  strncpy(g_assert_info.msg, msg, sizeof(g_assert_info.msg));
+  __asm("bkpt 2");
+}
+
+void two_true(void) {
+  MY_ASSERT(0, "Assert in `assert2.c::two_true`");
+}
+```
+
+The drawbacks of this method is that every message is stored in the resulting
+binary, and this takes up precious code space.
+
+Below, we use `objdump` to view the resulting strings in the binary. We can see
+that each assert string takes up ~32 bytes.
+
+```
+$ arm-none-eabi-objdump -s -j '.text' build/nrf52.elf
+ ...
+ 2710 63000000 41737365 72742069 6e206061  c...Assert in `a
+ 2720 73736572 74322e63 3a3a7477 6f5f7472  ssert2.c::two_tr
+ 2730 75650000 41737365 72742069 6e206061  ue..Assert in `a
+ 2740 73736572 74322e63 3a3a7477 6f5f6661  ssert2.c::two_fa
+ 2750 6c736500                             lse.
+```
+
+The longer the string, the more codespace that is used, so this would actually
+**discourage** the use of asserts, which is not ideal.
+
+#### Extracted Information
+
+We can look up the file and line number of the assert hit assuming all assert
+strings are unique.
+
+### File**path** and Line Number Assert (#3)
+
+The most common approach I've seen for assets is to include the filepath and
+line number into the assert macro, using the `__FILE__` and `__LINE__` defines
+provided by the compiler.
+
+```c
+#define MY_ASSERT(expr) \
+  do {                                          \
+    if (!(expr)) {                              \
+      my_assert(__FILE__, __LINE__);            \
+    }                                           \
+  } while (0)
+```
+
+```c
+void my_assert(const char *file, uint32_t line) {
+  strncpy(g_assert_info.file, file, sizeof(g_assert_info.file));
+  g_assert_info.line = line;
+  __asm("bkpt 3");
+}
+
+void two_true(void) {
+  MY_ASSERT(0);
+}
+```
+
+This approach works very well and gives the developer exactly the amount of
+context required to find the issue. However, the `__FILE__` macro includes the
+**entire** filepath of the file, and those can get quite lengthy. In my local
+build, `__FILE__` is 77 bytes!
+
+```
+(gdb) p __FILE__
+$1 = "/Users/tyler/dev/memfault/interrupt/example/debugging-asserts/impls/assert3.c"
+```
+
+Thankfully, the `__FILE__` string table entry is re-used between asserts in the
+same file, so the code size bloat primarily depends on the number of files with
+asserts within them. As a rough estimate, if we have asserts in 50 different
```

Comment:
It kinda feels like you need a transition word here since the first sentence sounds very positive and the second sounds very negative. Something like `Still, as a rough estimate, ...`? Not really sure 🤷‍♂

Reply:
Nah, you are right. I added:

`However, most projects have a non-trivial number of files` in between